### ISSUE TITLE
How can I use my GPU 😂

### ISSUE THREAD

**Issue by loveandbeepbrother** [guidance-ai/guidance#536]
**The bug**
I used the 0.15 version of guidance, and the compiled version of the current source code. But no matter how I adjust the value of the device parameter, (0, 'auto', 'gpu', etc.) 
every time the model is loaded in the memory and the following error is reported:
```python
ValueError: Found modules on cpu/disk. 
Using Exllama or Exllamav2 backend requires all the modules to be on GPU.You can deactivate exllama backend by setting `disable_exllama=True` in the quantization config object
```

I use the A30, Driver Version: 515.65.01 CUDA Version: 11.7,torch ==2.0.1
```python
import guidance
from transformers import AutoModelForCausalLM, AutoTokenizer
model_path = "Qwen-14B-Chat-Int4"

tokenizer = AutoTokenizer.from_pretrained(model_path, trust_remote_code=True)
guidance.llm = guidance.models.Transformers(model_path,tokenizer,device=0, trust_remote_code=True)
```
P.S: torch.cuda.is_available() = True
**System info (please complete the following information):**
 - OS (e.g. Ubuntu, Windows 11, Mac OS, etc.):ubuntu 18.04
 - Guidance Version (`guidance.__version__`):0.1.5\0.1.7

**Comment by slundberg** [comment]
Have you tried loading the same model with vanilla transformers? If that works but the guidance load fails let us know :)

**Comment by loveandbeepbrother** [comment]
> Have you tried loading the same model with vanilla transformers? If that works but the guidance load fails let us know :)  

yeah,I tried them in same notebook
```python
tokenizer = AutoTokenizer.from_pretrained(model_path, trust_remote_code=True)
model = AutoModelForCausalLM.from_pretrained(model_path, device_map="auto", trust_remote_code=True).eval()
model.generation_config = GenerationConfig.from_pretrained(model_path, trust_remote_code=True)
```

```python
import guidance
from transformers import AutoModelForCausalLM, AutoTokenizer
model_path = "Qwen-14B-Chat-Int4"

tokenizer = AutoTokenizer.from_pretrained(model_path, trust_remote_code=True)
guidance.llm = guidance.models.Transformers(model_path,tokenizer,device=0, trust_remote_code=True)
```
and only guidance failed😂

**Comment by ambroser53** [comment]
I had this issue as well, guidance doesn't make it very clear because they have a weird "`device`" flag that casts the model to that device after loading it automatically to cpu (which is wrong on so many levels). Saving grace is that guidance allows you to pass whatever you want to the transformers from_pretrained call within kwargs and what your supposed to use is `device_map`. I use `device_map={"":0}` which just always tries to fit the model on the first GPU on the machine. Also don't forget to set your `torch_dtype` to something like `bfloat16` because it'll often load in `float32` by default which is not only overkill but will not fit in most cards for inference.

**Comment by loveandbeepbrother** [comment]
> I had this issue as well, guidance doesn't make it very clear because they have a weird "`device`" flag that casts the model to that device after loading it automatically to cpu (which is wrong on so many levels). Saving grace is that guidance allows you to pass whatever you want to the transformers from_pretrained call within kwargs and what your supposed to use is `device_map`. I use `device_map={"":0}` which just always tries to fit the model on the first GPU on the machine. Also don't forget to set your `torch_dtype` to something like `bfloat16` because it'll often load in `float32` by default which is not only overkill but will not fit in most cards for inference.

IT JUST WORKS!!! Thanks a lot😀

**Comment by jeanibarz** [comment]
Have the same issue. I would prefear to use llama-cpp-python instead of ctransformers to load GGUF quantized models. Do you have some guidance ? I'm running guidance inside a devcontainer. The devcontainer is correctly configured: I have been able to run TTS on GPU on it. I tried the following:
- changed my docker base image, which I know was working fine with my GPU, from `nvidia/cuda:11.6.2-cudnn8-runtime-ubuntu20.04` to `nvidia/cuda:12.1.1-devel-ubuntu22.04`
- uninstalled llama-cpp-python and reinstall it with the following command:
`CMAKE_ARGS="-DLLAMA_OPENBLAS=on" FORCE_CMAKE=1 pip install llama-cpp-python --force-reinstall --upgrade --no-cache-dir`

But the model still load on my RAM and is executed on CPU. My devcontainer is using the following docker-compose.yml definition:
```
version: "3.8"

services:
  app:
    build:
      context: .
      dockerfile: .devcontainer/Dockerfile
      args:
        USER_UID: 1000
        USER_GID: 1000
    init: true
    volumes:
      - .:/workspace:cached
      - llm_models:/workspace/models:cached
    command: sleep infinity
    restart: no
    network_mode: host
    deploy:
      resources:
        reservations:
          devices:
            - driver: nvidia
              count: all
              capabilities: [gpu]
    shm_size: '24gb'
    environment:
      FORCE_CUDA: 1
      CUDA_VISIBLE_DEVICES: 0

volumes:
  llm_models:
    external: true
```

**Comment by jeanibarz** [comment]
Just solved my issue. Problem was coming from llama-cpp-python.
I made this script to troubleshoot my issue:
```python
from llama_cpp import Llama

llama_config = dict(
    model_path="/workspace/models/TheBloke/Mistral-7B-Instruct-v0.2-DARE-GGUF/mistral-7b-instruct-v0.2-dare.Q3_K_L.gguf",
    n_ctx=4096,
    n_gpu_layers=33,
    device_map={"": 0},
)

llm = Llama(**llama_config)
output = llm(
    "Q: Give very lengthy and detailed explanation of nuclear fusion. A: ",  # Prompt
    max_tokens=1024,
    stop=[
        "Q:",
    ],  # Stop generating just before the model would generate a new question
    echo=True,  # Echo the prompt back in the output
)  # Generate a completion, can also call create_completion
print(output)
```

Got 1024 completion tokens in 54s, i.e. ~20 tokens/sec (I had ~5 tokens/sec on the CPU). The console log and GPU memory also indicates GPU is correctly used.

**Solution that worked for me:** I had to use CUBLAS instead of OPENBLAS, i.e. I used this command for installing llama-cpp-python with GPU support:
```
CMAKE_ARGS="-DLLAMA_CUBLAS=ON" FORCE_CMAKE=1 pip install llama-cpp-python --force-reinstall --upgrade --no-cache-dir
```

instead of this one:
```
CMAKE_ARGS="-DLLAMA_OPENBLAS=on" FORCE_CMAKE=1 pip install llama-cpp-python --force-reinstall --upgrade --no-cache-dir
```

**Comment by cpcdoy** [comment]
GPU usage for the `Transformers` wrapper is fixed in #569.

As for `llama-cpp-python` it needs to be compiled with [`cuBLAS`](https://github.com/abetlen/llama-cpp-python#cublas) for Nvidia GPUs support and [`HipBLAS`](https://github.com/abetlen/llama-cpp-python#hipblas) for AMD GPUs support. You can also alternatively use [`CLBlas`](https://github.com/abetlen/llama-cpp-python#clblast) (not recommended for Nvidia GPUs since it offers less optimizations).

**Comment by slundberg** [comment]
I am going to close this since the fix from @cpcdoy in #569 addressed the main issue here.
